jpl.pl

```diff
@@ -4333,6 +4351,209 @@
     [ nl, '  ~q'-[H] ],
     dir_per_line(T).
 
-% Initialize JVM
+         /*******************************
+         *      EXCEPTION HANDLING      *
+         *******************************/
+
+% Cleartext messages that are carried by exceptions.
+% They are looked up by a meaningless atom (i.e. an "error code")
+% No effort has been made to collapse identicial messages ... yet.
+
+exception_msg(ea01,'1st arg must be bound to a classname, descriptor or object type').
```

Comment:
There are lots of error codes and man of them overlap in their text almost 90%. For example instantiation errors differ only on the argument number.

Would it be better to do:

```
exception_msg(instantiation_error, 1, [N], <concatenate N with " arg must be bound to l...')
```

The `1` is the first code number for type `instatiation_error` and we can add a list to pass data to build a particular message template (list will be empty in many cases).


I may be missing something.

Reply:
Error exceptions __must__ be of the shape `error(Formal, Context)` to make the tools work correctly.  Formal should as much as possible follow the ISO exceptions, but it is ok to make up jpl_*(...) terms.  Context is unbound or a term context(Location, Comment) where Location is either unbound or a predicate indicator.  Comment is just a comment string.   I think @dtonhofer follows these guidelines mostly (at least doesn't move away from them).

Reply:
If we go ahead with this update, we can group the errors in the next round and find better identifiers, I can look into that. (I still have the changes to the DCG pending in any case). 

I indeed follow the ISO-Standard (as does the original code of jpl.pl) with this exception:

On line 997, "system error" is thrown with arguments. However "system error" takes no arguments (ISO-Standard is really a bit wonky in the whole exception section and needs review) and probably shouldn't be thrown by user code, but by the Prolog Processor only.

So I changed 

`throw_system_error(Pred,array_element_type,T,eq03) `

to

`throw_illegal_state_error(Pred,array_element_type,T,eq03)` throwing the exception term `error(illegal_state_error,context(Pred,MsgText))`.

But that isn't ISO-Standard conforming either, and the open question is whether the exception term should pretend to look like an ISO-Standard exception term `error(Formal, Context)` or should affirm that it is jpl-specific: `jpl_error(Formal, Context)`. I would vote for the latter...

Reply:
1. I agree with doing a first round and then a second round to smooth a few corners, like grouping messages. 

2. I agree to use the new term `jpl_error/2`  for the special different case. In that case you just need to do the DCG for this new case right? I think Jan is also OK with that resolution as per his previous comment (and suggestion to add the DCG rule).

I am happy to progress on both, thanks for all this nice clean-up!

Reply:
Thank you. Actually, the DCG case is a bit more complex but we can discuss that in time.

Reply:
> whether the exception term should pretend to look like an ISO-Standard exception term `error(Formal, Context)` or should affirm that it is jpl-specific: `jpl_error(Formal, Context)`. I would vote for the latter...

No.  _all_ errors _must_ use error(Formal, Context) because that is what the development environment expects.  _Formal_ should be one of the ISO terms if (reasonably) appropriate.  If nothing is appropriate you may look in `boot/messages.pl` whether SWI-Prolog already defines something more appropriate or you may invent your own _Formal_.   In this case my vote would got for `jpl_state_error` with enough arguments to provide enough context for a good message.   So, __not__ _illegal_ `state_error`.  Look at the ISO terms.  There is no illegal_type_error, _bad_ something, etc.  It would be double, we are already talking about an _error_.

Reply:
Got it. I will propose something.

Reply:
After some consideration, I would just say we stay with the ISO "type error". Here is why:

- jpl_set_array/4, the predicate under consideration, already throws ISO "type error" on two places: 
   - when any of the _Prolog Datums_ to be punched into the _Java array_ does not have a corresponding _Java type_. 
   - when any of the _Java types_ derived from the _Prolog Datums_ does not match the _Java array's element type_.

Now, technically ISO "type error" is all about a type problem in the Prolog universe, so the above is an unexpected extension (one day one might switch to a dedicated jpl `Formal` indicating "datum type mismatch" and "datum array element type mismatch" instead).

The "illegal state" occurs when the indicated _Java array's element type_ is neither a class, nor an array, nor a primitive. But if we throw ISO "type error" in the cases above, we certainly can throw ISO "type error" in that case. 

So no specific jpl `Formal`, at least for now.

Reply:
Updated jpl.pl in that sense & pushed to repo.

Reply:
Sounds good, you have refactored it to be a `type_error`, makes sense. And that predicate will throw that error always.

With this fix, no need to add any DCG, right?

(And it is now clear we should **always** throw `error(.., ..)` terms.)

Reply:
Thanks Sebastian. But no, the DCG is for another problem, namely parsing the Java class names. I will grab my code and test cases and submit that one separately.

Reply:
OK, great! 

I may be missing  something David, but I thought you would indeed had to add a DCG rule  if you were to create a new Formal error, say `jpl_state_error`, of the form 

```prolog
prolog:error_message(jpl_state_error) --> ...
```

Since you are now using ISO `type_error`, then it is already handled by SWIPL via the rule 

```
prolog:message(error(Error, context(Stack, Message))) -->
...
```

in `library/prolog_stack.pl` (where I can see why Jan stated everything hsa to be `error/2`, for this rule to apply) which in term will use `translate_message/1` to translate `Error` and then `iso_message` to translate all ISO errors (like `type_error`.

That is my reconstruction so far... :-)

So, ready to merge then this improvement?

Reply:
Ah sorry. We are not on the same page: I was talking about the DCGs to parse/generate the Java classnames. I actually didn't know about the DCGs that generate the error texts at all. So no need for changes to the latter. More on the former ... soon.

Reply:
@JanWielemaker , can you merge into master or do you want me to do it? Maybe it is best you always merge important changes so we don't overlap.

While the changes are over the old repo structure, the merge would work perfectly and it is passing the [build check](https://github.com/SWI-Prolog/packages-jpl/pull/60/checks)

Reply:
Hold on. I just realized that the `java_in_prolog` test is failing one test:

```
64: Test command: /home/ssardina/git/soft/prolog/swipl-devel.git/build/src/swipl "-p" "foreign=:/home/ssardina/git/soft/prolog/swipl-devel.git/build/packages/plunit" "-f" "none" "--no-packs" "-s" "/home/ssardina/git/soft/prolog/swipl-devel.git/packages/jpl/test_jpl.pl" "-g" "test_jpl" "-t" "halt"
64: Test timeout computed to be: 9.99988e+06
64: % PL-Unit: jpl ......................................................
64: ERROR: /home/ssardina/git/soft/prolog/swipl-devel.git/packages/jpl/test_jpl.pl:767:
64: 	test set_field_bad_field_spec_1: wrong error
64: ERROR:     Expected: error(type_error(field_name,3.7),context(jpl_set/3,A))
64: ERROR:     Got:      error(existence_error(procedure,jpl:throw_type_error/3),context(jpl:jpl_set_static/4,B))
```

It is this test:

```prolog
test(
        set_field_bad_field_spec_1,
        [       setup((
                        BadFieldName = 3.7
                )),
                throws(
                        error(
                                type_error(field_name,BadFieldName),
                                context(jpl_set/3,_)
                        )
                )
        ]
) :-
    jpl_set('org.jpl7.Test', BadFieldName, a).
```

@dtonhofer it is expecting a `type_error` but it is getting a `existence_error`. Can you check this in your fix?

Reply:
Have found issue @dtonhofer , it was a missing `Pred` in one of the new `throw_type_error` calls, should be trivial to fix, thanks.